### DISCUSSION TITLE
When the data(parameter in am_recv_callback) will be freed or marked resuable?

### DISCUSSION THREAD

**Discussion by XCloudAI** [openucx/ucx#7012] [Q&A] [answered]
For active message API, there's below API will be called when receving the data from peer side:
https://github.com/openucx/ucx/blob/v1.10.1/test/apps/iodemo/ucx_wrapper.cc#L465
```
ucs_status_t UcxContext::am_recv_callback(void *arg, const void *header,
                                          size_t header_length,
                                          void *data, size_t length,
                                          const ucp_am_recv_param_t *param)
```
After the ```void *data``` has been copied, who's responsible for releasing the space or mark it could be resuable?
https://github.com/openucx/ucx/blob/v1.10.1/test/apps/iodemo/ucx_wrapper.cc#L643
```
    if (!(data_desc._param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV)) {
        memcpy(buffer, data_desc._data, length);
        (*callback)(UCS_OK);
        return true;
    }
```

**Comment by yosefe** [comment, marked as answer]
`UcxContext::am_recv_callback` function is part of io_demo test application, not UCX core library. For documentation of active messages API, please see https://github.com/openucx/ucx/blob/master/src/ucp/api/ucp_def.h#L669 - the "void *data" will be released according to the **return** value of the active message callback.

**Comment by XCloudAI** [comment]
@yosefe 
Take io_demo for example:
1. set active message callback here:
https://github.com/openucx/ucx/blob/v1.10.1/test/apps/iodemo/ucx_wrapper.cc#L142
```
    if (_use_am) {
        set_am_handler(am_recv_callback, this);
    } else {
        recv_io_message();
    }
```
2. the ```am_recv_callback``` will always return ```UCS_OK```
https://github.com/openucx/ucx/blob/v1.10.1/test/apps/iodemo/ucx_wrapper.cc#L490
```
    self->dispatch_am_message(conn, header, header_length, data_desc);

    return UCS_OK
```

@yosefe So, after ```am_recv_callback``` is called, the ```void *data``` will be released, right?
If so, is there some method to **cache** the ```void *data``` space, then the user could call some API to release it to UCX library.
The reason that I want to do this is because:
After ```am_recv_callback``` is retured, the ```void *data``` space will be released. So, I need to copy the data into one cache space first (assume its name is ```am_cache```). In some scenario, I need to trigger the other component to read the received data. So I copy the data from ```am_cache``` again. In such case, I do one time extra copy. If the ```void *data``` could be cached and controled by user to decied when releasing it to UCX, then It could save one time of copy operation.

**Comment by XCloudAI** [comment]
@yosefe What's about the usage of ```UCP_AM_RECV_ATTR_FLAG_DATA```? Could I use it to control that the ```void *data``` could serverd by user and release it to UCX later?

**Reply by yosefe** [reply]
if UCP_AM_RECV_ATTR_FLAG_DATA flag is present, that it's possible to return UCS_INPROGRESS from the active message handler, and release it later by ucp_am_data_release

**Comment by changchengx** [comment]
@XCloudAI 
|> The reason that I want to do this is because:

Could you show me the code in your project?

**Reply by XCloudAI** [reply]
Just school's homework. I've graduated these days.
